istioctl/pkg/tag/util.go

```diff
@@ -0,0 +1,168 @@
+		nsNames[i] = ns.Name
+	}
+	return nsNames, nil
+}
+
+// GetWebhookName extracts tag name from webhook object.
+func GetWebhookName(wh admit_v1.MutatingWebhookConfiguration) (string, error) {
+	if tagName, ok := wh.ObjectMeta.Labels[IstioTagLabel]; ok {
+		return tagName, nil
+	}
+	return "", fmt.Errorf("could not extract tag name from webhook")
+}
+
+// GetWebhookRevision extracts tag target revision from webhook object.
+func GetWebhookRevision(wh admit_v1.MutatingWebhookConfiguration) (string, error) {
+	if tagName, ok := wh.ObjectMeta.Labels[label.IoIstioRev.Name]; ok {
+		return tagName, nil
+	}
+	return "", fmt.Errorf("could not extract tag revision from webhook")
+}
+
+// DeleteTagWebhooks deletes the given webhooks.
+func DeleteTagWebhooks(ctx context.Context, client kubernetes.Interface, webhooks []admit_v1.MutatingWebhookConfiguration) error {
+	var result error
+	for _, wh := range webhooks {
+		result = multierror.Append(client.AdmissionregistrationV1().MutatingWebhookConfigurations().Delete(ctx, wh.Name, metav1.DeleteOptions{})).ErrorOrNil()
+	}
+	return result
+}
+
+var neverMatch = &metav1.LabelSelector{
+	MatchLabels: map[string]string{
+		"istio.io/deactivated": "never-match",
+	},
+}
+
+// DefaultRevisionExists checks whether there is an existing default revision, either implicit by virtue of having a
+// previous non-revisioned installation or an explicit default tag. Should be used in installer when deciding whether
+// to make an installation the default.
+func DefaultRevisionExists(ctx context.Context, client kubernetes.Interface) (bool, error) {
+	tagWhs, err := GetWebhooksWithTag(ctx, client, DefaultRevisionName)
+	if err != nil {
+		return false, err
+	}
+	whs, err := GetWebhooksWithRevision(ctx, client, DefaultRevisionName)
+	if err != nil {
+		return false, err
+	}
+	return len(tagWhs) > 0 || len(whs) > 0, nil
+}
+
+// DeactivateIstioInjectionWebhook deactivates the istio-injection webhook from the given MutatingWebhookConfiguration if exists.
+// used rather than just deleting the webhook since we want to keep it around after changing the default so user can later
+// switch back to it. This is a hack but it is meant to cover a corner case where a user wants to migrate from a non-revisioned
+// old version and then later decides to switch back to the old revision again.
+func DeactivateIstioInjectionWebhook(ctx context.Context, client kubernetes.Interface) error {
+	admit := client.AdmissionregistrationV1().MutatingWebhookConfigurations()
+	whs, err := GetWebhooksWithRevision(ctx, client, DefaultRevisionName)
+	if err != nil {
+		return err
+	}
+	if len(whs) == 0 {
+		// no revision with default, no action required.
+		return nil
+	}
+	if len(whs) > 1 {
+		return fmt.Errorf("expected a single webhook for default revision")
+	}
+	webhook := whs[0]
+	for i := range webhook.Webhooks {
+		wh := webhook.Webhooks[i]
+		// this is an abomination, but if this isn't a per-revision webhook, we want to make it ineffectual
```

Comment:
Is it possible to just keep the MutatingWebhookConfiguration with empty list of Webhooks?

Reply:
Yeah the reason I did it this way is we need to keep the injection URL around _somewhere_ so we can create new tags from the webhook. Right now the webhooks are the only place it's persisted.

Reply:
Ah, lame. Thats annoying. I don't know how often users are actually doing this but i was pretty confusing looking at `kubectl get mutatingwebhooks` and seeing a bunch, when the yaml shows disabled. In theory we could stick it in an annotation but I am not sure that is worth the complexity, since in some cases its not just a simple URL (port, cabundle, etc).

I guess the comment is accurate :slightly_smiling_face:

Reply:
I have no doubt that someone, somewhere will eventually come to rely on `"istio.io/deactivated": "never-match"` for injection and we'll have to keep it around :smile: